### PR TITLE
Fortis Cassandra Intregration

### PR DESCRIPTION
**Pull request by erikschlegel**

### CODE REVIEW

File: src/test/scala/com/microsoft/partnercatalyst/fortis/spark/sink/cassandra/CassandraIntegrationTestSpec.scala
Comment (jcjimenez) [inline src/test/scala/com/microsoft/partnercatalyst/fortis/spark/sink/cassandra/CassandraIntegrationTestSpec.scala]: the humanity! 😲

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/FortisEvent.scala
Comment (jcjimenez) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/FortisEvent.scala]: No biggie if we want to leave it - just want to confirm this is intentional or if we meant to remove `genders`.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/kafka/KafkaSink.scala
Comment (jcjimenez) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/kafka/KafkaSink.scala]: Similar question about `genders` here.

**Review by jcjimenez — APPROVED** [review]
LGTM with minor questions.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/FortisEvent.scala
Comment (erikschlegel) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/FortisEvent.scala]: Commenting out as this isnt quite ready for primetime just yet. This will be going out within the next two weeks.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/CassandraTest.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/CassandraTest.scala:18]: We use 2 space for indentation

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/CassandraTest.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/CassandraTest.scala]: Add return before this line

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/FortisEvent.scala
Comment (kevinhartman) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/dto/FortisEvent.scala]: Please add at least a "TODO:" above this so we don't lose it.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraConfig.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraConfig.scala]: Ideally, these should be on the`FortisSettings` object, but let's address that in a separate PR.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventSchema.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventSchema.scala:25]: Is this correct? I thought I remembered this being named "insertion_time" in the cql setup. If so, it should be "insertionTime" here.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventSchema.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventSchema.scala]: `getFeatures`

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala:29]: is `time` used?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: IMO the code would be easier to follow if you were to remove this method and call `.saveToCassandra` inside `writeFortisEvents`.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: It's weird to me for this to return something. Without reading the implementation, my assumption was that this would return the deduped events. Better to return nothing here and simply do the conversion to the event schema inside the `foreachRdd`.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: Under what condition would the session be null?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: Collecting these and having an in-memory set means they will be pulled from the workers to the driver. Your filter step below will then copy this set for every task to the workers. Ideally what you want is to do something like a `leftJoin` from the merged event identifier DF to the events DF (do events.toDF to get that from the RDD). Left join may be a method on dataframes you can just use.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala:88]: This would be better expressed as an `if else`

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: Indent just 2 spaces.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/aggregators/FortisAggregator.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/aggregators/FortisAggregator.scala:21]: Method names should be camel-cased.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/aggregators/FortisAggregator.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/aggregators/FortisAggregator.scala]: I get nervous when I see something called `flatMap` that's not a collection. Is there a better name that can be used here?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/dto/FortisRecords.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/dto/FortisRecords.scala]: Please fix indents.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: Indent this whole block

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: The `par` thing is actually probably fine, though I hope it's safe to submit spark tasks from non-spark threads.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: Is this space padding arbitrary? Can it be removed?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/aggregators/FortisAggregator.scala
Comment (kevinhartman) [inline src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/aggregators/FortisAggregator.scala:21]: These method names are a bit confusing. Method names should be actions unless they're pseudo lightweight properties (even then they should be camel-cased).  I'd revisit the `FortisAggregator` trait now or in a refactoring PR to iron out these names/signatures. It's unclear to me how these methods use their input parameters and what I'll get back from just reading the contract.

**Review by kevinhartman — APPROVED** [review]
Please fix the indentation issues before merging. We've been pretty strictly following the 2 space indent rule, and this PR has a lot of offenses there!

The biggest issue I see is around pairing up events with event ids from the batch (reading back from Cassandra), in which you're bringing everything to the driver. See comment in context for fix.

In general, this PR is too big to thoroughly review in one sitting. I expect we will need to iterate on some things introduced here.

When merging, *please squash commits*. This can be done with the merge button. There are a lot of uninteresting commits in the history of this PR that we should avoid checking in.

**Comment by kevinhartman** [comment]
Please **squash** commits for this merge (or squash individual commits into sets manually with git's interactive rebase if you're so inclined).
screen shot 2017-08-08 at 8 18 26 pm

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraConfig.scala
Comment (erikschlegel) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraConfig.scala]: Agree and added

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventSchema.scala
Comment (erikschlegel) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventSchema.scala:25]: yes, removed `_` across the board for all cassandra column names.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/CassandraTest.scala
Comment (erikschlegel) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/CassandraTest.scala]: Done

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (erikschlegel) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala:29]: figure we'd use it for logging to app insight. @Smarker ?

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (erikschlegel) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: Phew and Great! Thanks for confirming.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (erikschlegel) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: If the connection is dropped, which should never happen. I can remove the check is it should never happen.

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (erikschlegel) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: It shouldn't. I'll set the return type to `Unit`

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (erikschlegel) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: Done

File: src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala
Comment (erikschlegel) [reply src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala]: I hear you. Added the following ```private def fetchEventBatch(batchid: String, events: RDD[Event], session: SparkSession): Dataset[Event] = {
    import session.sqlContext.implicits._

    val addedEventsDF = session.sqlContext.read.format(CassandraFormat)
      .options(Map("keyspace" -> KeyspaceName, "table" -> TableEventBatches))
      .load()

    addedEventsDF.createOrReplaceTempView(TableEventBatches)
    val ds = session.sqlContext.sql(s"select eventid, pipelinekey from $TableEventBatches where batchid = '$batchid'")
    val eventsDS = events.toDF().as[Event]
    val filteredEvents = eventsDS.join(ds, Seq("eventid", "pipelinekey"))

    filteredEvents.as[Event]
  }```

**Pull request merged**
